File: avro-examples/java/TestSchema.java
import java.io.*;
import java.util.*;

import org.apache.avro.Schema;
import org.apache.avro.Schema.Type;
import org.apache.avro.Schema.Field;
import org.apache.avro.file.DataFileWriter;

public class TestSchema {
  public static Schema schemaFromCode() {
    List<Field> fields = new ArrayList<Field>();
    fields.add(new Field("field1", Schema.create(Type.STRING), null, null));
    fields.add(new Field("field2", Schema.create(Type.LONG), null, null));
    fields.add(new Field("field3", Schema.createArray(Schema.create(Type.STRING)), null, null));
    fields.add(new Field("field4", Schema.createMap(Schema.create(Type.INT)), null, null));

    Schema schema = Schema.createRecord("recordName", "Record Doc String", "recordNS", false);
    schema.setFields(fields);

    schema.addProp("Property1", "Value1");
    schema.addProp("Property2", "Value2");

    return(schema);
  }

  public static void writeSchema (File file, String json) throws IOException {
    FileWriter writer = new FileWriter(file);
    writer.write(json, 0, json.length());
    writer.close();
  }

  public static void main(String[] args) throws IOException {
    Schema s1 = schemaFromCode();
    String jsonSchema = s1.toString(true);

    File file = new File("test-schema.json");
    writeSchema(file, jsonSchema);

    Schema s2 = Schema.parse(file);
    assert(s1.equals(s2));

    System.out.println(jsonSchema);
  }
}

File: avro-examples/c/test-schema.c
/*
 * gcc -Wall test-schema.c -o test-schema -I/usr/include/ -L/usr/local/lib/ -lavro
 * LD_LIBRARY_PATH=/usr/local/lib ./test-schema
 */
#include <assert.h>

#include <string.h>
#include <stdlib.h>
#include <stdio.h>

#include <avro.h>

static avro_schema_t __schema_from_code (void) {
    avro_schema_t field_sub_schema;
    avro_schema_t field_schema;
    avro_schema_t schema;

    schema = avro_schema_record("recordName", "recordNS");

    field_schema = avro_schema_string();
    avro_schema_record_field_append(schema, "field1", field_schema);
    avro_schema_decref(field_schema);

    field_schema = avro_schema_long();
    avro_schema_record_field_append(schema, "field2", field_schema);
    avro_schema_decref(field_schema);

    field_sub_schema = avro_schema_string();
    field_schema = avro_schema_array(field_sub_schema);
    avro_schema_decref(field_sub_schema);
    avro_schema_record_field_append(schema, "field3", field_schema);
    avro_schema_decref(field_schema);

    field_sub_schema = avro_schema_int();
    field_schema = avro_schema_map(field_sub_schema);
    avro_schema_decref(field_sub_schema);
    avro_schema_record_field_append(schema, "field4", field_schema);
    avro_schema_decref(field_schema);

    return(schema);
}

int main (int argc, char **argv) {
    avro_schema_error_t error;
    avro_schema_t s1, s2;
    avro_writer_t writer;
    char json[1024];

    s1 = __schema_from_code();

    writer = avro_writer_memory(json, sizeof(json));
    avro_schema_to_json(s1, writer);
    avro_writer_free(writer);

    avro_schema_from_json(json, strlen(json), &s2, &error);
    assert(avro_schema_equal(s1, s2));

    printf("%s\n", json);

    avro_schema_decref(s1);
    avro_schema_decref(s2);

    return(0);
}

